src/algebra/polynomial/group_ring_action.lean

```diff
@@ -15,6 +15,11 @@ This file contains instances and definitions relating `mul_semiring_action` to `
 
 variables (M : Type*) [monoid M]
 
+/-- These instances describe an incompatible scalar action, and make the lemmas in this file
+hard to state. -/
+local attribute [-instance]
+  polynomial.has_scalar polynomial.distrib_mul_action
```

Comment:
I am a little bit worried by this comment. Can you explain a little bit more what is going on here? (If we have two different scalar actions hanging around, this might create many confusing situations down the road).

Reply:
The comment in the other file captures it better:
```
/-- In this file we are interested only in `polynomial.mul_semiring_action` which transfers the
action to each variable, and not these instances which transfer the action to each coefficient. -/
```
Essentially, there are two different incompatible `distrib_mul_action` structures that can be put on a polynomial.
Previously their requirements were different:
1. `polynomial.mul_semiring_action : mul_semiring_action A (polynomial B)` requires `mul_semiring_action A B`
2. `polynomial.has_scalar : has_scalar A (polynomial B)` requires `module A B`
In practice these two typeclasses are mutually exclusive; `mul_semiring_action.smul_one : ∀ (a : A), a • (1 : B) = 1` and `module.zero_smul : ∀ (b : B), (0 : A) • b = 0` together would imply that `B` is the trivial ring, so the action would be the same

This PR observes that `polynomial.has_scalar` has stronger typeclasses than it needed, and relaxes them. Unfortunately, that means the requirements for 2 are now a subset of the requirements for 1, meaning that 1 is always found first. I can see a few ways to resolve this:

* Remove `[instance]` from `polynomial.has_scalar` and `polynomial.distrib_mul_action`, and put turning them back on in a locale. This locale would be open for stating lemmas like `smul_monomial` with maximal generality, but would not be open in the tiny number of files that use the `mul_semiring_action`.
* Leave those instances on by default (they're the "obvious" ones), and create a locale that disables them purely for the sake of `mul_semiring_action`. That would give a nice place to put a slightly longer comment than I have here, shared by the two files that need it.
* Remove `mul_semiring_action` entirely in place of using a type alias, `mul_semiring_action (map_action A) (polynomial B)`.
* Do not generalize the `polynomial` instances for now. I only really wanted this for `finsupp`, and figured I may as well generalize things downstream too.

Reply:
I'm very confused here. What is the mathematical content of

> `polynomial.mul_semiring_action : mul_semiring_action A (polynomial B)` requires `mul_semiring_action A B`

?

Reply:
I think it means "If the is a distributive action such that for `a : A` `b1 b2 : B`, `a \smul (b1 * b2) = (a \smul b1) * (a \smul v2)`, then this action in inherited by polynomials by defining `a \smul (b * X^n) as `a^n \smul (b*X^n)`"

Reply:
Is there an example to think about where one cares about a `mul_semiring_action` on polynomials (... or indeed at all?)

Maybe I'm just experiencing lack of imagination, but I don't see why `polynomial.mul_semiring_action` merits existence.

Reply:
I don't know for sure why it does either, but I didn't want to think about it too much in this PR.

Reply:
Seems to have been added in #2586, maybe @kckennylau has an example?

Reply:
https://github.com/leanprover-community/mathlib/blob/31ba155fbe0f78c7091df64fe087418203f964a0/src/field_theory/fixed.lean#L59

Reply:
@kckennylau, what are your thoughts on the `attribute [-instance]` in this PR? The problem is that any type which has an action suitable for use in the domain of the polynomial evaluation function (`polynomial.mul_semiring_action`) is also suitable for use in the codomain (`polynomial.has_scalar`)

Reply:
I see. Mathematically this is saying that:

1. `polynomial.mul_semiring_action`: if a monoid A acts on a ring B, then A also acts on B[X];
2. `polynomial.has_scalar`: if A is a ring and B is an A-module, then B[X] is also an A-vector space.

There is no conflict currently, until we realize that the scalar multiplication in 2 can be defined without any module hypothesis. Am I summarizing the issue correctly?

But I still don't understand. Lean should just try 1 and see that it fails and then move on to 2, right?

Reply:
> There is no conflict currently, until we realize that the scalar multiplication in 2 can be defined without any module hypothesis. Am I summarizing the issue correctly?

Correct, this PR is making the change that follows frmo that realization

> But I still don't understand. Lean should just try 1 and see that it fails and then move on to 2, right?

Lean would only do this if `mul_semiring_action.to_distrib_mul_action` had higher priority than `polynomial.distrib_mul_action`. But that would be very weird - "forgetful" instances like this are usually priority 100, and making `polynomial.distrib_mul_action` have an even lower priority would be very strange.

Reply:
> 1. `polynomial.mul_semiring_action`: if a monoid A acts on a ring B, then A also acts on B[X];

This meaning of "act" is not what `has_scalar` is used to mean pretty much anywhere but in a very specific case (actions by subgroups of algebra automorphisms over fields on polynomials of that field, which follow from `intermediate_field.subgroup_action` onwards).

Reply:
In the interest of making progress:

> Do not generalize the polynomial instances for now. I only really wanted this for finsupp, and figured I may as well generalize things downstream too.

I've done this in #7819

Reply:
Sorry for the confusion, my conclusions were wrong here. I've removed these `[-instance]` lines and fixed the diamond. Thanks for making me think about this more!